input/postgres/schema.go

```diff
@@ -27,28 +28,13 @@ func CollectAllSchemas(server *state.Server, collectionOpts state.CollectionOpts
 	ps.Functions = []state.PostgresFunction{}
 
 	for _, dbName := range schemaDbNames {
-		schemaConnection, err := EstablishConnection(server, logger, collectionOpts, dbName)
+		psNext, databaseOid, err := collectOneSchema(server, collectionOpts, logger, ps, dbName, ts.Version)
 		if err != nil {
-			logger.PrintVerbose("Failed to connect to database %s to retrieve schema: %s", dbName, err)
+			logger.PrintVerbose("Failed to collect schema metadata for database %s: %s", dbName, err)
 			continue
 		}
-
-		databaseOid, err := CurrentDatabaseOid(schemaConnection)
-		if err != nil {
-			logger.PrintError("Error getting OID of database %s", dbName)
-			schemaConnection.Close()
-			continue
-		}
-
-		ps.SchemaStats[databaseOid] = &state.SchemaStats{
-			RelationStats: make(state.PostgresRelationStatsMap),
-			IndexStats:    make(state.PostgresIndexStatsMap),
-		}
-
-		ps = collectSchemaData(collectionOpts, logger, schemaConnection, ps, databaseOid, dbName, ts.Version, server.Config.IgnoreSchemaRegexp)
+		ps = psNext
 		ts.DatabaseOidsWithLocalCatalog = append(ts.DatabaseOidsWithLocalCatalog, databaseOid)
-
-		schemaConnection.Close()
```

Comment:
Extracted most of the loop to a function to make it easier to ensure the connection is closed in all paths by using `defer`.